### PR TITLE
sdp: egpio: backends functionality improvements

### PR DESCRIPTION
**Pull request by jaz1-nordic**
- Add MBOX improvements to ensuring that data has been read by the remote core.
- Add general structure code improvements to header files.

### CODE REVIEW

File: applications/sdp/gpio/src/backend/backend.h [hunk 1]
```diff
@@ -9,7 +9,8 @@
 
 #include <drivers/gpio/nrfe_gpio.h>
 
-#if !defined(CONFIG_GPIO_NRFE_EGPIO_BACKEND_ICMSG) && !defined(CONFIG_GPIO_NRFE_EGPIO_BACKEND_MBOX)
+#if !IS_ENABLED(CONFIG_GPIO_NRFE_EGPIO_BACKEND_ICMSG) &&                                           \
+	!IS_ENABLED(CONFIG_GPIO_NRFE_EGPIO_BACKEND_MBOX)
```
Comment (magp-nordic) [inline applications/sdp/gpio/src/backend/backend.h:13]: This is already modified in icbmsg PR: https://github.com/nrfconnect/sdk-nrf/pull/17321/files#diff-7fffc679701a822b8559ff5a9d885be0e635b124f4615690170ba7c39c1b177eR12-R14

File: drivers/gpio/gpio_nrfe_icmsg.c [hunk 2]
```diff
@@ -11,7 +11,7 @@
 #include "gpio_nrfe.h"
 
 #if defined(CONFIG_MULTITHREADING)
-K_SEM_DEFINE(bound_sem, 0, 1);
+static K_SEM_DEFINE(bound_sem, 0, 1);
```
[the remark this answers is not in the document]
Comment (jaz1-nordic) [reply drivers/gpio/gpio_nrfe_icmsg.c:14]: Commit message changed.

**Review by magp-nordic — APPROVED** [review]

**Comment by jaz1-nordic** [comment]
@nrfconnect/ncs-co-drivers Could you please take a look?

File: include/sdp/nrfe_common.h [hunk 3]
```diff
@@ -4,22 +4,29 @@
  * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
  */
 
-#ifndef NRFE_COMMON_H__
-#define NRFE_COMMON_H__
+#ifndef SDP_NRFE_COMMON_H__
+#define SDP_NRFE_COMMON_H__
 
-#include <stdatomic.h>
+#include <zephyr/sys/atomic.h>
```
Comment (masz-nordic) [inline include/sdp/nrfe_common.h:10]: How does that affect code size? Are there any dependencies to using zephyr's atomic subsys?
Can you show assembly diff?

Comment (masz-nordic) [reply include/sdp/nrfe_common.h:10]: Both `sys/atomic` and `stdatomic` use GCC builtins. So should be fine.

**Review by masz-nordic — APPROVED** [review]

**Comment by rlubos** [comment]
@nrfconnect/ncs-co-drivers Please review

File: applications/sdp/gpio/src/backend/backend.h [hunk 4]
```diff
@@ -9,9 +9,9 @@
 
 #include <drivers/gpio/nrfe_gpio.h>
 
-#if !defined(CONFIG_GPIO_NRFE_EGPIO_BACKEND_ICMSG) && \
-	!defined(CONFIG_GPIO_NRFE_EGPIO_BACKEND_MBOX) && \
-	!defined(CONFIG_GPIO_NRFE_EGPIO_BACKEND_ICBMSG)
+#if !IS_ENABLED(CONFIG_GPIO_NRFE_EGPIO_BACKEND_ICMSG) && \
+	!IS_ENABLED(CONFIG_GPIO_NRFE_EGPIO_BACKEND_MBOX) && \
+	!IS_ENABLED(CONFIG_GPIO_NRFE_EGPIO_BACKEND_ICBMSG)
```
Comment (anangl) [inline applications/sdp/gpio/src/backend/backend.h:14]: Why?

**Comment by nordic-segl** [comment]
@nrfconnect/ncs-co-drivers Please review
This is needed for NCS 2.8.0.

File: applications/sdp/gpio/src/backend/backend.h (hunk 4 above)
Comment (jaz1-nordic) [reply applications/sdp/gpio/src/backend/backend.h:14]: https://github.com/nrfconnect/sdk-nrf/pull/16592#discussion_r1784363820

Comment (anangl) [reply applications/sdp/gpio/src/backend/backend.h:14]: > [#16592 (comment)](https://github.com/nrfconnect/sdk-nrf/pull/16592#discussion_r1784363820)

Symbols generated for Kconfig bool options are either defined as 1 or undefined. That's why using `IS_ENABLED()` in `#if` statements [is discouraged](https://github.com/zephyrproject-rtos/zephyr/blob/afce5329f772a4306af699c7a06efddf93cadca0/include/zephyr/sys/util_macro.h#L133-L134) (see also https://github.com/zephyrproject-rtos/zephyr/pull/53200).
But reverting this is probably not worth another CI run.

**Review by anangl — APPROVED** [review]

File: applications/sdp/gpio/src/backend/backend.h (hunk 4 above)
Comment (jaz1-nordic) [reply applications/sdp/gpio/src/backend/backend.h:14]: I will do it during the next PR.

**Pull request merged**